File: pset6/credit/credit.py
# Get the users input
credit_card = input("Number: ")


# Store the lenght of the credit card
lenght = len(credit_card)


# Store the first and second digit
first_digit = int(credit_card[0])
second_digit = int(credit_card[1])

check_sum = 0

# Calculate the Luhn's Algorithm
for i, number in enumerate(credit_card):
    if i % 2 == 0:
        check_sum += int(credit_card[lenght - i - 1])
    else:
        even = int(credit_card[lenght - i - 1]) * 2
        if even < 10:
            check_sum += even
        else:
            even = str(even)
            check_sum += int(even[0])
            check_sum += int(even[1])
        
# if valid, decide which credit card is
if check_sum != 20:
    print("INVALID")
    
    
# American Express    
if lenght == 15 and first_digit == 3 and (second_digit == 4 or second_digit == 7):
    print("AMEX")
    

# Mastercard
if lenght == 16 and first_digit == 5 and (second_digit > 0 and second_digit < 6):
    print("MASTERCARD")
    

# Visa
if (lenght == 13 or lenght == 16) and first_digit == 4:
    print("VISA")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    //Ask for the credit card number.
    long credit_card;

    do
    {
        credit_card = get_long("Enter your credit card: \n");
    }
    while (credit_card < 0);

    // Variable Declaration
    int digit = 0; //Takes a single digit each time of the credit card.

    int counter = 0; //Counts how many passes in thw while loop.

    int even_digit = 0; //Saves when the even digit has two digits.

    int sum = 0; // Stores the check sum.

    int first_digit = 0; //Stores the first digit.

    int second_digit = 0; //Stores the second digit.
    
    int check = 0; //save second digit

    while (credit_card != 0) //Start analyzing the credit card.
    {
        
        digit = (credit_card % 10);
        
        if ((credit_card / 100) == 0 && (check == 0))
        {
            second_digit = digit;
            check = 1; 
        }
        
        credit_card = (credit_card / 10);
        counter++;
        
        if ((credit_card % 10) == 0)
        {
            first_digit = digit;
        }
        
        if ((counter % 2) > 0) // Take the odd digits.
        {
            sum = (sum + digit);
        }

        if ((counter % 2) == 0) //Take the even digits.
        {
            digit *= 2;

            if (digit >= 10)
            {
                for (int j = 0; j < 2; j++)
                {
                    even_digit = (digit % 10);
                    digit = (digit / 10);
                    sum = (sum + even_digit);
                }
            }
            else
            {
                sum = (sum + digit);
            }
        }
    }

    // Final evaluations for the result
    if ((sum % 10) == 0)
    {
        if ((first_digit == 4) && (counter == 13 || counter == 16))
        {
            printf("VISA\n");
        }
        else if ((first_digit == 3) && (second_digit == 4 || second_digit == 7) && (counter == 15))
        {
            printf("AMEX\n");
        }
        else if ((first_digit == 5) && (counter == 16))
        {
            if (second_digit == 1 || second_digit == 2 || second_digit == 3 || second_digit == 4 || second_digit == 5)
            {
                printf("MASTERCARD\n");      
            }
            else
            {
                printf("INVALID\n"); 
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
